### PR TITLE
[SPARK-35015][PYTHON] Add missing type hints of pyspark.sql.*

### PR DESCRIPTION
**Pull request by Yikun**
### What changes were proposed in this pull request?

This patch adds the missing type hints of pyspark.sql.*:
- pyspark.sql.column._to_java_column
- pyspark.sql.column._to_seq
- pyspark.sql.column._create_column_from_literal
- pyspark.sql.types._parse_datatype_string

### Why are the changes needed?

To cleanup these error: 
python/pyspark/pandas/frame.py:6486: error: Module 'pyspark.sql.types' has no attribute '_parse_datatype_string'
python/pyspark/pandas/spark/functions.py:22: error: Module 'pyspark.sql.column' has no attribute '_to_java_column'
python/pyspark/pandas/spark/functions.py:22: error: Module 'pyspark.sql.column' has no attribute '_to_seq'
python/pyspark/pandas/spark/functions.py:22: error: Module 'pyspark.sql.column' has no attribute '_create_column_from_literal'

Related: https://issues.apache.org/jira/browse/SPARK-34941

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
1. Disable the ignore_error on mypy.ini and run mypy
https://github.com/apache/spark/blob/ee7bf7d9628384548d09ac20e30992dd705c20f4/python/mypy.ini#L132
2. lint-python passed.

### CODE REVIEW

File: python/pyspark/sql/column.pyi [hunk 1]
```diff
@@ -116,3 +123,11 @@ class Column:
     def __nonzero__(self) -> None: ...
     def __bool__(self) -> None: ...
     def contains(self, item: Any) -> Column: ...
+
+def _to_java_column(col: ColumnOrName) -> JavaObject: ...
```
Comment (Yikun) [inline python/pyspark/sql/column.pyi:127]: All missing type hints were [internal use](https://pep8.org/#descriptive-naming-styles) methods, but it still be used by other modules: https://github.com/apache/spark/blob/364d1eaf10f51c357f507325557fb076140ced2c/python/pyspark/pandas/spark/functions.py#L22

I think we should move these methods into pyspark.sql.utils or somewhere in future.

**Comment by Yikun** [comment]
cc @zero323

**Comment by zero323** [comment]
Thank you for your proposal @Yikun.

In general we decided to avoid annotating private and internal code (in fact, we even dropped annotations for some modules during original migration).

Could you explain what is the use case for adding these?

**Comment by Yikun** [comment]
> Could you explain what is the use case for adding these?

The annotating missing warning was found when I cleaunup the mypy error of spark pandas, see more in [SPARK-34941](https://issues.apache.org/jira/browse/SPARK-34941).

[python/pyspark/pandas/frame.py:6486](https://github.com/apache/spark/blob/214a46aa88cd682874584dc407ad130a30761884/python/pyspark/pandas/frame.py#L6486): error: Module 'pyspark.sql.types' has no attribute '_parse_datatype_string'
[python/pyspark/pandas/spark/functions.py:22](https://github.com/apache/spark/blob/214a46aa88cd682874584dc407ad130a30761884/python/pyspark/pandas/spark/functions.py#L22): error: Module 'pyspark.sql.column' has no attribute '_to_java_column'
[python/pyspark/pandas/spark/functions.py:22](https://github.com/apache/spark/blob/214a46aa88cd682874584dc407ad130a30761884/python/pyspark/pandas/spark/functions.py#L22): error: Module 'pyspark.sql.column' has no attribute '_to_seq'
[python/pyspark/pandas/spark/functions.py:22](https://github.com/apache/spark/blob/214a46aa88cd682874584dc407ad130a30761884/python/pyspark/pandas/spark/functions.py#L22): error: Module 'pyspark.sql.column' has no attribute

**Comment by zero323** [comment]
Thank you for your response. I think we need to discuss this further.  Personally, I'd lean towards using type ignore within newly added pandas modules. We could also switch to stub files to make Pandas consistent with the rest of the code.

Any thoughts, @HyukjinKwon?

**Comment by HyukjinKwon** [comment]
Yeah, I would prefer to ignore them for now too. It's planned to be revisited at SPARK-34941.

**Comment by HyukjinKwon** [comment]
cc @ueshin, @itholic, @xinrong-databricks FYI

**Comment by Yikun** [comment]
@zero323 @HyukjinKwon OK, thanks, let's close it first, I will help on when it's ready to start.

**Pull request closed without merging**